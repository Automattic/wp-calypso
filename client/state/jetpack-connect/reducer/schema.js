/** @format */
export const jetpackConnectSessionsSchema = {
	type: 'object',
	additionalProperties: false,
	patternProperties: {
		'^.+$': {
			type: 'object',
			required: [ 'timestamp' ],
			properties: {
				timestamp: { type: 'number' },
				flowType: { type: 'string' },
			},
			additionalProperties: false,
		},
	},
};

export const jetpackConnectAuthorizeSchema = {
	type: 'object',
	additionalProperties: false,
	patternProperties: {
		'^.+$': {
			type: 'object',
			required: [ 'queryObject', 'timestamp' ],
			properties: {
				authorizationCode: { type: 'string ' },
				authorizeError: { type: 'boolean' },
				authorizeSuccess: { type: 'boolean' },
				autoAuthorize: { type: 'boolean' },
				isAuthorizing: { type: 'boolean' },
				isRedirectingToWpAdmin: { type: 'boolean' },
				plansUrl: { type: 'string' },
				queryObject: {
					type: 'object',
					required: [
						'_wp_nonce',
						'client_id',
						'redirect_uri',
						'scope',
						'secret',
						'site',
						'state',
					],
					properties: {
						_wp_nonce: { type: 'string' },
						client_id: { type: 'string' },
						from: { type: 'string' },
						home_url: { type: 'string' },
						jp_version: { type: 'string' },
						new_user_started_connection: { type: 'boolean' },
						redirect_after_auth: { type: 'string' },
						redirect_uri: { type: 'string' },
						scope: { type: 'string' },
						secret: { type: 'string' },
						site: { type: 'string' },
						site_icon: { type: 'string' },
						site_url: { type: 'string' },
						state: { type: 'string' },
					},
					additionalProperties: false,
				},
				timestamp: { type: 'number' },
				siteReceived: { type: 'boolean' },
			},
			additionalProperties: false,
		},
	},
};

export const jetpackAuthAttemptsSchema = {
	type: 'object',
	additionalProperties: false,
	patternProperties: {
		'^.+$': {
			type: 'object',
			required: [ 'attempt', 'timestamp' ],
			attempt: { type: 'number' },
			timestamp: { type: 'number' },
		},
	},
};

export const jetpackConnectSelectedPlansSchema = {
	type: 'object',
	additionalProperties: false,
	patternProperties: {
		'^.+$': { type: [ 'string', 'null' ] },
	},
};
