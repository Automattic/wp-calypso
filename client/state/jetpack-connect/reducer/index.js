/** @format */
/**
 * Internal dependencies
 */
import jetpackAuthAttempts from './jetpack-auth-attempts';
import jetpackConnectAuthorize from './jetpack-connect-authorize';
import jetpackConnectSelectedPlans from './jetpack-connect-selected-plans';
import jetpackConnectSessions from './jetpack-connect-sessions';
import jetpackConnectSite from './jetpack-connect-site';
import jetpackSSO from './jetpack-sso';
import { combineReducers } from 'state/utils';

export default combineReducers( {
	jetpackConnectSite,
	jetpackSSO,
	jetpackConnectAuthorize,
	jetpackConnectSessions,
	jetpackConnectSelectedPlans,
	jetpackAuthAttempts,
} );
