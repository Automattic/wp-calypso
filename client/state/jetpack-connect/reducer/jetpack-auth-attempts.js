/**
 * Internal dependencies
 */
import { AUTH_ATTEMPS_TTL } from '../constants';
import { isStale } from '../utils';
import {
	JETPACK_CONNECT_COMPLETE_FLOW,
	JETPACK_CONNECT_RETRY_AUTH,
} from 'calypso/state/jetpack-connect/action-types';
import { jetpackAuthAttemptsSchema } from './schema';
import { keyedReducer, withSchemaValidation } from 'calypso/state/utils';

export function authAttempts( state = undefined, { type, attemptNumber } ) {
	switch ( type ) {
		case JETPACK_CONNECT_RETRY_AUTH:
			if ( ! state || isStale( state.timestamp, AUTH_ATTEMPS_TTL ) ) {
				return {
					attempt: 0,
					timestamp: Date.now(),
				};
			}
			return {
				attempt: attemptNumber,
				timestamp: state.timestamp,
			};

		case JETPACK_CONNECT_COMPLETE_FLOW:
			return undefined;
	}
	return state;
}

export const reducer = withSchemaValidation(
	jetpackAuthAttemptsSchema,
	keyedReducer( 'slug', authAttempts )
);

export default reducer;
