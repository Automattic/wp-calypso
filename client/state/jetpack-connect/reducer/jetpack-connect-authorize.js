/** @format */
/**
 * External dependencis
 */
import { includes, isEmpty, omit } from 'lodash';

/**
 * Internal dependencies
 */
import {
	JETPACK_CONNECT_COMPLETE_FLOW,
	JETPACK_CONNECT_QUERY_SET,
	JETPACK_CONNECT_AUTHORIZE,
	JETPACK_CONNECT_AUTHORIZE_LOGIN_COMPLETE,
	JETPACK_CONNECT_AUTHORIZE_RECEIVE,
	JETPACK_CONNECT_AUTHORIZE_RECEIVE_SITE_LIST,
	JETPACK_CONNECT_CREATE_ACCOUNT,
	JETPACK_CONNECT_CREATE_ACCOUNT_RECEIVE,
	JETPACK_CONNECT_REDIRECT_WP_ADMIN,
	JETPACK_CONNECT_REDIRECT_XMLRPC_ERROR_FALLBACK_URL,
	JETPACK_CONNECT_USER_ALREADY_CONNECTED,
	SITE_REQUEST_FAILURE,
	SERIALIZE,
	DESERIALIZE,
} from 'state/action-types';
import { isStale } from '../utils';
import { JETPACK_CONNECT_AUTHORIZE_TTL } from '../constants';

function buildDefaultAuthorizeState() {
	return {
		queryObject: {},
		isAuthorizing: false,
		authorizeSuccess: false,
		authorizeError: false,
		timestamp: Date.now(),
		userAlreadyConnected: false,
		autoAuthorize: false,
	};
}

export default function jetpackConnectAuthorize( state = {}, action ) {
	switch ( action.type ) {
		case JETPACK_CONNECT_AUTHORIZE:
			return Object.assign( {}, omit( state, 'userData', 'bearerToken' ), {
				isAuthorizing: true,
				authorizeSuccess: false,
				authorizeError: false,
				isRedirectingToWpAdmin: false,
				autoAuthorize: false,
			} );
		case JETPACK_CONNECT_AUTHORIZE_RECEIVE:
			if ( isEmpty( action.error ) && action.data ) {
				const { plans_url } = action.data;
				return Object.assign( {}, state, {
					authorizeError: false,
					authorizeSuccess: true,
					autoAuthorize: false,
					plansUrl: plans_url,
					siteReceived: false,
				} );
			}
			return Object.assign( {}, state, {
				isAuthorizing: false,
				authorizeError: action.error,
				authorizeSuccess: false,
				autoAuthorize: false,
			} );
		case JETPACK_CONNECT_AUTHORIZE_LOGIN_COMPLETE:
			return Object.assign( {}, state, { authorizationCode: action.data.code } );
		case JETPACK_CONNECT_AUTHORIZE_RECEIVE_SITE_LIST:
			const updateQueryObject = omit( state.queryObject, '_wp_nonce', 'secret', 'scope' );
			return Object.assign( {}, omit( state, 'queryObject' ), {
				siteReceived: true,
				isAuthorizing: false,
				queryObject: updateQueryObject,
			} );
		case JETPACK_CONNECT_QUERY_SET:
			const queryObject = Object.assign( {}, action.queryObject );
			const shouldAutoAuthorize = includes(
				[ 'woocommerce-services-auto-authorize', 'woocommerce-setup-wizard' ],
				queryObject.from
			);
			const additionalQueryArgs = shouldAutoAuthorize ? { autoAuthorize: true } : {};
			return Object.assign(
				{},
				buildDefaultAuthorizeState(),
				{ queryObject },
				additionalQueryArgs
			);
		case JETPACK_CONNECT_CREATE_ACCOUNT:
			return Object.assign( {}, state, {
				isAuthorizing: true,
				authorizeSuccess: false,
				authorizeError: false,
				autoAuthorize: true,
			} );
		case JETPACK_CONNECT_CREATE_ACCOUNT_RECEIVE:
			if ( ! isEmpty( action.error ) ) {
				return Object.assign( {}, state, {
					isAuthorizing: false,
					authorizeSuccess: false,
					authorizeError: true,
					autoAuthorize: false,
				} );
			}
			return Object.assign( {}, state, {
				isAuthorizing: true,
				authorizeSuccess: false,
				authorizeError: false,
				autoAuthorize: true,
				userData: action.userData,
				bearerToken: action.data.bearer_token,
			} );
		case SITE_REQUEST_FAILURE:
			if (
				state.queryObject &&
				state.queryObject.client_id &&
				parseInt( state.queryObject.client_id ) === action.siteId
			) {
				return Object.assign( {}, state, { clientNotResponding: true } );
			}
			return state;
		case JETPACK_CONNECT_USER_ALREADY_CONNECTED:
			return Object.assign( {}, state, { userAlreadyConnected: true } );
		case JETPACK_CONNECT_REDIRECT_XMLRPC_ERROR_FALLBACK_URL:
			return Object.assign( {}, state, { isRedirectingToWpAdmin: true } );
		case JETPACK_CONNECT_REDIRECT_WP_ADMIN:
			return Object.assign( {}, state, { isRedirectingToWpAdmin: true } );
		case JETPACK_CONNECT_COMPLETE_FLOW:
			return {};
		case DESERIALIZE:
			return ! isStale( state.timestamp, JETPACK_CONNECT_AUTHORIZE_TTL ) ? state : {};
		case SERIALIZE:
			return state;
	}
	return state;
}
jetpackConnectAuthorize.hasCustomPersistence = true;
