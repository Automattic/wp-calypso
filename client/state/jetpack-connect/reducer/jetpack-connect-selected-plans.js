/** @format */
/**
 * Internal dependencies
 */
import { JETPACK_CONNECT_CHECK_URL, JETPACK_CONNECT_COMPLETE_FLOW } from 'state/action-types';
import { jetpackConnectSelectedPlansSchema } from './schema';

export default function jetpackConnectSelectedPlans( state = {}, action ) {
	switch ( action.type ) {
		case JETPACK_CONNECT_CHECK_URL:
			return { '*': state[ '*' ] };
		case JETPACK_CONNECT_COMPLETE_FLOW:
			return {};
	}
	return state;
}
jetpackConnectSelectedPlans.schema = jetpackConnectSelectedPlansSchema;
