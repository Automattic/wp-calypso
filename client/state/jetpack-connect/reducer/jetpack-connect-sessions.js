/** @format */
/**
 * External dependencis
 */
import { pickBy } from 'lodash';

/**
 * Internal dependencies
 */
import { DESERIALIZE, JETPACK_CONNECT_CHECK_URL, SERIALIZE } from 'state/action-types';
import { isValidStateWithSchema } from 'state/utils';
import { jetpackConnectSessionsSchema } from './schema';
import { isStale } from '../utils';
import { urlToSlug } from 'lib/url';

function buildUrlSessionObj( url, flowType ) {
	const slug = urlToSlug( url );
	const sessionValue = {
		timestamp: Date.now(),
		flowType: flowType || '',
	};
	return { [ slug ]: sessionValue };
}

export default function jetpackConnectSessions( state = {}, action ) {
	switch ( action.type ) {
		case JETPACK_CONNECT_CHECK_URL:
			return Object.assign( {}, state, buildUrlSessionObj( action.url, action.flowType ) );
		case DESERIALIZE:
			if ( isValidStateWithSchema( state, jetpackConnectSessionsSchema ) ) {
				return pickBy( state, session => {
					return ! isStale( session.timestamp );
				} );
			}
			return {};
		case SERIALIZE:
			return state;
	}
	return state;
}
jetpackConnectSessions.hasCustomPersistence = true;
