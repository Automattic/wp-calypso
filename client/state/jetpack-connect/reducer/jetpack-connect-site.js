/** @format */
/**
 * Internal dependencies
 */
import {
	JETPACK_CONNECT_CHECK_URL,
	JETPACK_CONNECT_CHECK_URL_RECEIVE,
	JETPACK_CONNECT_DISMISS_URL_STATUS,
	JETPACK_CONNECT_CONFIRM_JETPACK_STATUS,
	JETPACK_CONNECT_COMPLETE_FLOW,
	JETPACK_CONNECT_REDIRECT,
} from 'state/action-types';

export default function jetpackConnectSite( state = {}, action ) {
	const defaultState = {
		url: null,
		isFetching: false,
		isFetched: false,
		isDismissed: false,
		installConfirmedByUser: null,
		data: {},
	};
	switch ( action.type ) {
		case JETPACK_CONNECT_CHECK_URL:
			return Object.assign( {}, defaultState, {
				url: action.url,
				isFetching: true,
				isFetched: false,
				isDismissed: false,
				installConfirmedByUser: null,
				data: {},
			} );
		case JETPACK_CONNECT_CHECK_URL_RECEIVE:
			if ( action.url === state.url ) {
				return Object.assign( {}, state, {
					isFetching: false,
					isFetched: true,
					data: action.data,
				} );
			}
			return state;
		case JETPACK_CONNECT_DISMISS_URL_STATUS:
			if ( action.url === state.url ) {
				return Object.assign( {}, state, { installConfirmedByUser: null, isDismissed: true } );
			}
			return state;
		case JETPACK_CONNECT_REDIRECT:
			if ( action.url === state.url ) {
				return Object.assign( {}, state, { isRedirecting: true } );
			}
			return state;
		case JETPACK_CONNECT_CONFIRM_JETPACK_STATUS:
			return Object.assign( {}, state, { installConfirmedByUser: action.status } );
		case JETPACK_CONNECT_COMPLETE_FLOW:
			return {};
	}
	return state;
}
