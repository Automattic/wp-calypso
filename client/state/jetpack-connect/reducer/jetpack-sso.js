/** @format */
/**
 * Internal dependencies
 */
import {
	JETPACK_CONNECT_SSO_AUTHORIZE_REQUEST,
	JETPACK_CONNECT_SSO_AUTHORIZE_SUCCESS,
	JETPACK_CONNECT_SSO_AUTHORIZE_ERROR,
	JETPACK_CONNECT_SSO_VALIDATION_REQUEST,
	JETPACK_CONNECT_SSO_VALIDATION_SUCCESS,
	JETPACK_CONNECT_SSO_VALIDATION_ERROR,
} from 'client/state/action-types';

export default function jetpackSSO( state = {}, action ) {
	switch ( action.type ) {
		case JETPACK_CONNECT_SSO_VALIDATION_REQUEST:
			return Object.assign( {}, state, { isValidating: true } );
		case JETPACK_CONNECT_SSO_VALIDATION_SUCCESS:
			return Object.assign( {}, state, {
				isValidating: false,
				validationError: false,
				nonceValid: action.success,
				blogDetails: action.blogDetails,
				sharedDetails: action.sharedDetails,
			} );
		case JETPACK_CONNECT_SSO_VALIDATION_ERROR:
			return Object.assign( {}, state, {
				isValidating: false,
				validationError: action.error,
				nonceValid: false,
			} );
		case JETPACK_CONNECT_SSO_AUTHORIZE_REQUEST:
			return Object.assign( {}, state, { isAuthorizing: true } );
		case JETPACK_CONNECT_SSO_AUTHORIZE_SUCCESS:
			return Object.assign( {}, state, {
				isAuthorizing: false,
				authorizationError: false,
				ssoUrl: action.ssoUrl,
			} );
		case JETPACK_CONNECT_SSO_AUTHORIZE_ERROR:
			return Object.assign( {}, state, {
				isAuthorizing: false,
				authorizationError: action.error,
				ssoUrl: false,
			} );
	}
	return state;
}
