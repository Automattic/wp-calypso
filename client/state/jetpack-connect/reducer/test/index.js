/** @format */
/**
 * Internal dependencies
 */
import reducer from '..';

describe( 'reducer', () => {
	test( 'should export expected reducer keys', () => {
		const initialState = reducer( undefined, {} );
		expect( initialState ).toMatchObject( {
			jetpackConnectSite: expect.anything(),
			jetpackConnectAuthorize: expect.anything(),
			jetpackConnectSessions: expect.anything(),
			jetpackSSO: expect.anything(),
			jetpackConnectSelectedPlans: expect.anything(),
			jetpackAuthAttempts: expect.anything(),
		} );
		expect( initialState ).toMatchSnapshot();
	} );
} );
