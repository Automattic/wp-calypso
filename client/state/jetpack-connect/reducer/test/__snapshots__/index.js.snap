// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer should export expected reducer keys 1`] = `
Object {
  "jetpackAuthAttempts": Object {},
  "jetpackConnectAuthorize": Object {},
  "jetpackConnectSessions": Object {},
  "jetpackConnectSite": Object {},
  "jetpackSSO": Object {},
}
`;
